pygfx/renderers/wgpu/textrender.py

```diff
@@ -0,0 +1,158 @@
+    suggested_render_mask = 3  # todo: needs work
+
+    # Put it together!
+    return [
+        {
+            "suggested_render_mask": suggested_render_mask,
+            "render_shader": shader,
+            "primitive_topology": wgpu.PrimitiveTopology.triangle_list,
+            "indices": (range(n), range(1)),
+            "vertex_buffers": {},
+            "bindings0": bindings,
+        }
+    ]
+
+
+class TextShader(WorldObjectShader):
+    def get_code(self):
+        return (
+            self.get_definitions()
+            + self.common_functions()
+            + self.vertex_shader()
+            + self.fragment_shader()
+        )
+
+    def vertex_shader(self):
+        return """
+
+        struct VertexInput {
+            @builtin(vertex_index) vertex_index : u32,
+        };
+
+
+        @stage(vertex)
+        fn vs_main(in: VertexInput) -> Varyings {
+
+            let index = i32(in.vertex_index);
+            let i0 = index / 6;
+            let sub_index = index % 6;
+
+            let glyph_pos = load_s_positions(i0);
+
+            let screen_factor = u_stdinfo.logical_size.xy / 2.0;
+
+            var deltas = array<vec2<f32>, 6>(
+                vec2<f32>(-1.0, -1.0),
+                vec2<f32>(-1.0,  1.0),
+                vec2<f32>( 1.0, -1.0),
+                vec2<f32>(-1.0,  1.0),
+                vec2<f32>( 1.0, -1.0),
+                vec2<f32>( 1.0,  1.0),
+            );
+            let size = 12.0;  // todo: where to get the size?
+            let aa_margin = 1.0;
+            let point_coord = deltas[sub_index] * (size + aa_margin);
+
+            $$ if screen_space
+
+                // We take the object's pos (model pos is origin), move to NDC, and apply the
+                // glyph-positioning in logical screen coords.
+
+                let raw_pos = vec3<f32>(0.0, 0.0, 0.0);
+                let world_pos = u_wobject.world_transform * vec4<f32>(raw_pos, 1.0);
+                let ndc_pos = u_stdinfo.projection_transform * u_stdinfo.cam_transform * world_pos;
+                let delta_ndc = (glyph_pos.xy + point_coord) / screen_factor;
+
+            $$ else
+
+                // We take the glyph positions as model pos, move to world and then NDC.
+
+                let raw_pos = glyph_pos + vec3<f32>(point_coord, 0.0);
+                let world_pos = u_wobject.world_transform * vec4<f32>(raw_pos, 1.0);
+                let ndc_pos = u_stdinfo.projection_transform * u_stdinfo.cam_transform * world_pos;
+                let delta_ndc = vec2<f32>(0.0, 0.0);
+
+            $$ endif
+
+            var varyings: Varyings;
+            varyings.position = vec4<f32>(ndc_pos.xy + delta_ndc * ndc_pos.w, ndc_pos.zw);
+            varyings.world_pos = vec3<f32>(world_pos.xyz / world_pos.w);
+            varyings.pointcoord = vec2<f32>(point_coord);
+            varyings.size = f32(size);
+
+            // Picking
+            varyings.pick_idx = u32(i0);
```

Comment:
Is this going to support individual glyph picking? I didn't expect that sort of awesomeness 🤠

Reply:
Yes. Even the location within the glyph :D   But note that glyph != character. Multiple characters can become a single glyph in the shaping step, so we'd need to account for that.

Reply:
💡 got it!